src/cargo/core/resolver/features.rs

```diff
@@ -732,10 +788,44 @@ impl<'a, 'cfg> FeatureResolver<'a, 'cfg> {
                         }
                         true
                     })
-                    .map(|dep| {
-                        let dep_for_host = self.track_for_host
-                            && (for_host || dep.is_build() || self.is_proc_macro(dep_id));
-                        (dep, dep_for_host)
+                    .flat_map(|dep| {
```

Comment:
This block of code is really critical to how the `FeaturesFor` tracking works.  However, it is kinda long and hard to follow (I could handle the old 3-line closure, but the new code is challenging). Can you sprinkle around some comments to try to make it clear what this is doing?  I'm not sure what to say, but perhaps start with a high-level explanation of what its purpose is?

Reply:
Thanks for pointing it out! When looking at the portion of code for a moment I was afraid I wouldn't get it together anymore. Fortunately, memory returned while pulling it apart to the point where I could make sense of it and think it's probably correct according to my knowledge.

I hope the _maybe too wordy_ annotations in 040e6195d4c788a58c644069bdeada711e103967 can make it clearer and serve as a basis for suggestions on correctness and general fit.